src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -428,31 +453,25 @@ protected override Task AppendProjectStateDiagnosticsAsync(Project project, Docu
 
             private async Task<CompilationWithAnalyzers> GetCompilationWithAnalyzersAsync(Project project, bool concurrentAnalysis, bool reportSuppressedDiagnostics, CancellationToken cancellationToken)
             {
-                CompilationWithAnalyzers compilationWithAnalyzers;
-                lock (_compilationWithAnalyzersMap)
-                {
-                    if (_compilationWithAnalyzersMap.TryGetValue(project.Id, out compilationWithAnalyzers))
-                    {
-                        return compilationWithAnalyzers;
-                    }
-                }
-
-                if (project.SupportsCompilation)
+                if (_compilationWithAnalyzers == null && project.SupportsCompilation)
                 {
                     var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
-                    compilationWithAnalyzers = Owner.GetCompilationWithAnalyzers(project, compilation, concurrentAnalysis, reportSuppressedDiagnostics);
-                }
-                else
-                {
-                    compilationWithAnalyzers = null;
+                    var analyzers = GetAnalyzers(project);
+
+                    _compilationWithAnalyzers = Owner.GetCompilationWithAnalyzers(project, analyzers, compilation, concurrentAnalysis, reportSuppressedDiagnostics);
                 }
 
-                lock (_compilationWithAnalyzersMap)
+                return _compilationWithAnalyzers;
+            }
+
+            protected IEnumerable<DiagnosticAnalyzer> GetAnalyzers(Project project)
+            {
+                if (_analyzers == null)
                 {
-                    _compilationWithAnalyzersMap[project.Id] = compilationWithAnalyzers;
+                    _analyzers = Owner._stateManager.GetOrCreateAnalyzers(project);
```

Comment:
not sure what you meant by it is not thread safe. _stateManager.GetOrCreateAnalyzers(project) is deterministic API. same input always give back same output. when there is a race (which I dont think there can be due to same reason above), we might run GetOrCreateAnalyzers multiple time, but that should be it.